### PR TITLE
feat(executor): Basic in-application-memory `DynamicFilterExecutor`, pass tpch q11, q22 e2e tests 

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Basic in-application-memory `DynamicFilterExecutor`, passes e2e tests 

Next PR: unit tests

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Release note

Add support for non-equal-condition joins (only scalar comparisons)

## Refer to a related PR or issue link (optional)
Tracking issue: https://github.com/singularity-data/risingwave/issues/3419

### CODE REVIEW

**Review by BowenXiao1999 — APPROVED** [review]
Rest LGTM

File: src/stream/src/executor/dynamic_filter.rs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        //
+        // TODO: It could be potentially expensive memory-wise to store `HashSet`.
+        //       If I'm not wrong, the memory overhead is a backing Vec of size 4
+        //       (See: https://github.com/rust-lang/hashbrown/pull/162)
+        //       + some byte-per-entry metadata. Well, `Row` is on heap anyway...
+        //       It could be preferred to find a way to do prefix range scans on the left key and
+        //       storing as `BTreeSet<(ScalarImpl, Row)>`.
+        //       We could solve it if `ScalarImpl` had a successor/predecessor function.
+        //
+        //       Probably, using a custom comparator function on a custom datatype
```
Comment (jon-chuang) [inline src/stream/src/executor/dynamic_filter.rs:97]: Actually, I will do this in this PR

File: src/stream/src/executor/dynamic_filter.rs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+        let mut stream_chunk_builder =
+            StreamChunkBuilder::new(PROCESSING_WINDOW_SIZE, &self.schema.data_types(), 0, 0)
+                .map_err(StreamExecutorError::eval_error)?;
+
+        #[for_await]
+        for msg in aligned_stream {
+            match msg? {
+                AlignedMessage::Left(chunk) => {
+                    // TODO: refactor into fn: `apply_batch`
```
Comment (jon-chuang) [inline src/stream/src/executor/dynamic_filter.rs:119]: Should reduce the per-function complexity. Will do this.

File: src/stream/src/executor/dynamic_filter.rs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+                    // Alternatively, the behaviour can be to flatten the deletion of
+                    // `current_epoch_value` into a NULL represented by a `None: Datum`
+                    if !last_is_insert {
+                        return Err(anyhow!("RHS updates should always end with inserts").into());
+                    }
+                }
+                AlignedMessage::Barrier(barrier) => {
+                    // Flush the difference between the `prev_value` and `current_value`
+                    // TODO: refactor into fn `flush_range`
```
Comment (jon-chuang) [inline src/stream/src/executor/dynamic_filter.rs:245]: Should reduce the per-function complexity. Will do this.

File: src/stream/src/from_proto/dynamic_filter.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        let condition = build_from_prost(prost_condition)?;
+        let comparator = prost_condition.get_expr_type()?;
+        if !matches!(
+            comparator,
+            GreaterThan | GreaterThanOrEqual | LessThan | LessThanOrEqual
+        ) {
+            return Err(internal_error(
+                "`DynamicFilterExecutor` only supports comparators:\
+                GreaterThan | GreaterThanOrEqual | LessThan | LessThanOrEqual",
+            ));
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply src/stream/src/from_proto/dynamic_filter.rs:51]: Actually, ensure only works with anyhow::Error, and there is no type cast. So ensure cannot be used.

File: src/stream/src/executor/dynamic_filter.rs [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+                    // Reuse the logic from `FilterExecutor`
+                    let chunk = chunk.compact()?; // Is this unnecessary work?
+                    let (data_chunk, ops) = chunk.into_parts();
+
+                    let mut new_ops = Vec::with_capacity(ops.len());
+                    let mut new_visibility = BitmapBuilder::with_capacity(ops.len());
+                    let mut last_res = false;
+
+                    for (row, op) in data_chunk.rows().zip_eq(ops.iter()) {
+                        // TODO: convert this to a batch operation?
```
Comment (jon-chuang) [inline src/stream/src/executor/dynamic_filter.rs:129]: Do this.

File: src/stream/src/from_proto/dynamic_filter.rs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            // keyspace_r: Keyspace::table_root(store, &right_table_id),
+            actor_id: params.actor_id as u64,
+            metrics: params.executor_stats,
+        };
+
+        DynamicFilterExecutorDispatcher::dispatch_by_kind(kind, args)
+    }
+}
+
+struct DynamicFilterExecutorDispatcher; //<S: StateStore>(PhantomData<S>);
```
Comment (fuyufjh) [inline src/stream/src/from_proto/dynamic_filter.rs:93]: Why need a `DynamicFilterExecutorDispatcher` here? I think DynamicFilter doesn't need a hash table

File: src/stream/src/executor/dynamic_filter.rs (hunk 1 above)
Comment (jon-chuang) [reply src/stream/src/executor/dynamic_filter.rs:97]: Basically, what we want is a multimap. But there is currently no efficient implementation for such a multimap.

**Review by fuyufjh — APPROVED** [review]
Generally LGTM

**Comment by fuyufjh** [comment]
related https://github.com/singularity-data/risingwave/issues/1380

**Comment by jon-chuang** [comment]
Seems like test failures are unrelated...?

**Pull request merged**